doc/HOWTO/documenting-functions-and-macros.md

```diff
@@ -0,0 +1,175 @@
+diff --git a/include/openssl/bio.h.in b/include/openssl/bio.h.in
+index c70185db34..4ddea2f96b 100644
+--- a/include/openssl/bio.h.in
++++ b/include/openssl/bio.h.in
+@@ -485,6 +485,9 @@ struct bio_dgram_sctp_prinfo {
+ #define BIO_set_dgram_dest(b, addr)   BIO_ctrl(b, BIO_CTRL_DGRAM_SET_PEER, 0, addr)
+ #define BIO_get_dgram_dest(b, addr)   BIO_ctrl(b, BIO_CTRL_DGRAM_GET_PEER, 0, addr)
+
+ +int BIO_set_dgram_foo(BIO* b, int foo);
+ +#define BIO_set_dgram_bar(b, bar) BIO_ctrl(b, BIO_CTRL_DGRAM_SET_ADDR, 0, bar)
+ +
+  /*
+   * name is cast to lose const, but might be better to route through a
+   * function so we can do it safely
+```
+
+If you run `make doc-nits`, you might be surprised that it only
+complains about the undocumented macro, not the function:
+
+```
+~/src/openssl/master$ make doc-nits
+
+/usr/bin/perl ./util/find-doc-nits -c -n -l -e
+include/openssl/bio.h: macro BIO_set_dgram_bar(3) undocumented
+# 1 macros undocumented (count is approximate)
+make: *** [Makefile:3833: doc-nits] Error 1
+```
+
+The explanation for this is that one important step is still missing,
+it needs to be done first: you need to run
+
+```
+make update
+```
+
+which triggers a scan of the public headers for new API functions.
+
+All new functions will be added to either `util/libcrypto.num`
+or `util/libssl.num`.
+Those files store the information about the symbols which need
+to be exported from the shared library resp. DLL.
+Among other stuff, they contain the ordinal numbers for
+the [module definition file](https://docs.microsoft.com/en-us/cpp/build/exporting-from-a-dll-using-def-files) of the Windows DLL,
+which is the reason for the `.num` extension.
+
+After running `make update`, you can use `git diff` to check the outcome:
+
+```
+~/src/openssl/master/doc/HOWTO$ git diff
+diff --git a/util/libcrypto.num b/util/libcrypto.num
+index 394f454732..fc3c67313a 100644
+--- a/util/libcrypto.num
++++ b/util/libcrypto.num
+@@ -5437,3 +5437,4 @@ BN_signed_bn2native                     ? 3_1_0   EXIST::FUNCTION:
+ ASYNC_set_mem_functions                 ?  3_1_0   EXIST::FUNCTION:
+ ASYNC_get_mem_functions                 ?  3_1_0   EXIST::FUNCTION:
+ BIO_ADDR_dup                            ?  3_1_0   EXIST::FUNCTION:SOCK
++BIO_set_dgram_foo                       ?  3_1_0   EXIST::FUNCTION:
+```
+
+The changes need to be committed, ideally as a separate commit:
+
+```
+git commit -a -m "make update"
+```
+
+which has the advantage that it can easily be discarded when it
+becomes necessary to rerun `make update`.
+
+Finally, we reached the point where `make doc-nits` complains about
+both symbols:
+
+```
+msp@msppc:~/src/openssl/master$ make doc-nits
+/usr/bin/perl ./util/find-doc-nits -c -n -l -e
+crypto: function BIO_set_dgram_foo(3) undocumented
+# 1 libcrypto names are not documented
+include/openssl/bio.h: macro BIO_set_dgram_bar(3) undocumented
+# 1 macros undocumented (count is approximate)
+make: *** [Makefile:3833: doc-nits] Error 1
+```
```

Comment:
````suggestion
```console
$ make doc-nits
/usr/bin/perl ./util/find-doc-nits -c -n -l -e
crypto: function BIO_set_dgram_foo(3) undocumented
# 1 libcrypto names are not documented
include/openssl/bio.h: macro BIO_set_dgram_bar(3) undocumented
# 1 macros undocumented (count is approximate)
make: *** [Makefile:3833: doc-nits] Error 1
```
````